tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -389,3 +393,185 @@ def finalize(self, state):
+    self.shape = shape
+    self.num_chains = num_chains
+    if dtype is tf.int64:
+      dtype = tf.float64
+    elif dtype.is_integer:
+      dtype = tf.float32
+    self.dtype = dtype
+
+  def initialize(self):
+    """Initializes an empty `RunningPotentialScaleReductionState`.
+
+    Returns:
+      state: `RunningPotentialScaleReductionState` representing a stream
+        of no inputs.
+    """
+    def _initialize_for_one_state(num_chains, shape):
+      """Initializes a running variance state for one group of Markov chains."""
+      var_stream = RunningVariance(shape, self.dtype)
+      return [var_stream.initialize()
+              for _ in range(num_chains)]
+
+    chain_var = nest.map_structure_up_to(
+        self.num_chains,
+        _initialize_for_one_state,
+        self.num_chains,
+        self.shape,
+        check_types=False
+    )
+    return RunningPotentialScaleReductionState(chain_var)
+
+  def update(self, state, new_sample, chain_axis=0, chunk_axis=None):
+    """Update the `RunningPotentialScaleReductionState` with a new sample.
+
+    Args:
+      state: `RunningPotentialScaleReductionState` that represents the
+        current state of running statistics.
+      new_sample: Incoming `Tensor` sample or (possibly nested) collection of
+        `Tensor`s with shape and dtype compatible with those used to form the
+        `RunningPotentialScaleReductionState`.
+      chain_axis: The sample axis that indexes into independent Markov chains
+        samples. For batched computation, this can either be a scalar value that
+        represents the chain axis across all rhat calculations, or a
+        structure that identically mimics `self.num_chains`.
+      chunk_axis: If chunking is desired, this is an integer that specifies the
+        axis with chunked samples. For individual samples, set this to `None`.
+        By default, samples are not chunked (`axis` is None). For batched
+        computation, this can either be a scalar value or `None` that
+        represents chunking semantics across all rhat calculations, or a
+        structure that identically mimics `self.num_chains`.
+
+    Returns:
+      state: `RunningPotentialScaleReductionState` with updated calculations.
+    """
+    var_stream = RunningVariance(self.shape, self.dtype)
+    chunk_axis = nest_util.broadcast_structure(
+        self.num_chains, chunk_axis
+    )
+    chain_axis = nest_util.broadcast_structure(
+        self.num_chains, chain_axis)
+
+    def _update_for_one_state(
+        num_chains, chain_var, new_sample, chain_axis, chunk_axis):
+      """Updates the running variance for one group of Markov chains."""
+      sample_rank = ps.rank(new_sample)
+      if sample_rank >= 2:
+        # make the axis denoting independent chains the leading dimension
+        perm = [chain_axis] + list(
+            range(0, chain_axis)) + list(
+                range(chain_axis + 1, sample_rank))
+        new_sample = tf.transpose(new_sample, perm)
```

Comment:
What weakness of the underlying API makes this transposition necessary?

Reply:
that would be having a `chain_axis` argument -- we don't know which dimension after the leading in the sample actually represents the chain axis. We could simply streamline everything at make the incoming samples conform to a certain shape, like in `tfp.mcmc.potential_scale_reduction` and the reducer if you think this is more annoying than convenient (i.e. code complexity doesn't really make sense for a relatively unhelpful utility)